### PR TITLE
[SPARK-40536][CONNECT] Make Spark Connect port configurable

### PR DESCRIPTION
**Pull request by amaliujia**
### What changes were proposed in this pull request?

Add `Connect` config and one connect gRPC config keys: `spark.connect.grpc.binding.port` as Int type.

### Why are the changes needed?

Currently Spark Connect gRPC port is hardcoded and we can make it configurable. 

### Does this PR introduce _any_ user-facing change?

No
### How was this patch tested?

Existing UT

### CODE REVIEW

**Comment by amaliujia** [comment]
@HyukjinKwon @cloud-fan @grundprinzip

File: core/src/main/scala/org/apache/spark/internal/config/Connect.scala [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.internal.config
+
+private[spark] object Connect {
```
Comment (grundprinzip) [inline core/src/main/scala/org/apache/spark/internal/config/Connect.scala:20]: As Spark connect is build as plugin, all of the configuration should ideally be not located in core. Is there a way we can move this to the connect module?

Comment (amaliujia) [reply core/src/main/scala/org/apache/spark/internal/config/Connect.scala:20]: I agree. let's see other reviewer's suggestions on this.

File: core/src/main/scala/org/apache/spark/internal/config/Connect.scala [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+  private[spark] val CONNECT_GRPC_DEBUG_MODE =
+    ConfigBuilder("spark.connect.grpc.debug.enabled")
+      .version("3.4.0")
+      .booleanConf
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/internal/config/Connect.scala:25]: can we add docs?

File: core/src/main/scala/org/apache/spark/internal/config/Connect.scala [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+      .createOptional
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/internal/config/Connect.scala:26]: why is it optional? Seems like it defaults to `true`

File: core/src/main/scala/org/apache/spark/internal/config/Connect.scala (hunk 1 above)
Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/internal/config/Connect.scala:20]: Yeah, it should better be placed within `connect` if possible. How do we use this plugin? I suspect the jar should be provided anyway (?)

File: core/src/main/scala/org/apache/spark/internal/config/Connect.scala [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.internal.config
+
+private[spark] object Connect {
+
+  private[spark] val CONNECT_GRPC_DEBUG_MODE =
+    ConfigBuilder("spark.connect.grpc.debug.enabled")
```
Comment (zhengruifeng) [inline core/src/main/scala/org/apache/spark/internal/config/Connect.scala:23]: it is not used now?

File: core/src/main/scala/org/apache/spark/internal/config/Connect.scala (hunk 1 above)
Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/internal/config/Connect.scala:20]: Does Spark Connect module depend on Spark Core? If yes, we can move this object to Spark Connect module.

Comment (amaliujia) [reply core/src/main/scala/org/apache/spark/internal/config/Connect.scala:20]: Yes it depends on core. Let me move this config and also address other comments.

File: core/src/main/scala/org/apache/spark/internal/config/Connect.scala (hunk 4 above)
Comment (amaliujia) [reply core/src/main/scala/org/apache/spark/internal/config/Connect.scala:23]:
```
  def startGRPCService(): Unit = {
    val debugMode = SparkEnv.get.conf.getBoolean("spark.connect.grpc.debug.enabled", true)
```

This is being called in the grpc service. I was thinking given that this is already define so probably just add this config as well.

File: core/src/main/scala/org/apache/spark/internal/config/Connect.scala (hunk 1 above)
Comment (amaliujia) [reply core/src/main/scala/org/apache/spark/internal/config/Connect.scala:20]: > Yeah, it should better be placed within `connect` if possible. How do we use this plugin? I suspect the jar should be provided anyway (?)

I believe there is a way to decide if the plugin is loaded and there is a default way of if this is loaded or not. Let me check that.

Comment (amaliujia) [reply core/src/main/scala/org/apache/spark/internal/config/Connect.scala]: Moved to connect module.

File: core/src/main/scala/org/apache/spark/internal/config/Connect.scala (hunk 4 above)
Comment (amaliujia) [reply core/src/main/scala/org/apache/spark/internal/config/Connect.scala]: It's used already but I guess this flag and what is tries to enable worth extra work and PR. 

I deleted this config to make this PR only focus on one thing.

File: core/src/main/scala/org/apache/spark/internal/config/Connect.scala (hunk 2 above)
Comment (amaliujia) [reply core/src/main/scala/org/apache/spark/internal/config/Connect.scala]: I deleted this config to make this PR only focus on one thing, which is the configurable. 

This config and what it tries to enable worth a different effort to fully finalize.

File: connect/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectService.scala [hunk 5]
```diff
@@ -189,7 +189,7 @@ object SparkConnectService {
    */
   def startGRPCService(): Unit = {
     val debugMode = SparkEnv.get.conf.getBoolean("spark.connect.grpc.debug.enabled", true)
-    val port = 15002
+    val port = SparkEnv.get.conf.getInt("spark.connect.grpc.binding.port", 15002)
```
Comment (cloud-fan) [inline connect/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectService.scala:192]: shall we call `SparkEnv.get.conf.get(CONNECT_GRPC_BINDING_PORT)`?

Comment (amaliujia) [reply connect/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectService.scala]: Makes sense! Done!

File: connect/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectService.scala [hunk 6]
```diff
@@ -185,11 +186,10 @@ object SparkConnectService {
   /**
    * Starts the GRPC Serivce.
    *
-   * TODO(SPARK-40536) Make port number configurable.
    */
   def startGRPCService(): Unit = {
     val debugMode = SparkEnv.get.conf.getBoolean("spark.connect.grpc.debug.enabled", true)
-    val port = 15002
+    val port = SparkEnv.get.conf.getInt(CONNECT_GRPC_BINDING_PORT.key, 15002)
```
Comment (cloud-fan) [inline connect/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectService.scala:192]: This means we are duplicating the default value `15002` in 2 places. `SparkConf.get` is a better API to use
```
  def get[T](entry: ConfigEntry[T]): T = {
    recordConfigEntryAccessDuringWarmUp[T](entry, false, this) {
      entry.readFrom(reader)
    }
  }
```

Comment (amaliujia) [reply connect/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectService.scala]: Yes this is what I confused on the `getInt` API which always ask for a default value, and the get, instead, return a string.

The difference of this get is based on the Entry parameter, which was ignored when I searched the API.

Done.

**Review by cloud-fan — APPROVED** [review]

**Review by grundprinzip — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**